File: day02/hs/Main.hs
#!/usr/bin/env runhaskell

import           Data.Function
import           Data.Functor
import qualified Data.Text      as T
import           Debug.Trace    (trace)

data CubeSet = CubeSet !Int !Int !Int
  deriving (Show, Eq, Bounded)

instance Semigroup CubeSet where
  (CubeSet n i j) <> (CubeSet x y z) = CubeSet (n+x) (i+y) (j+z)

instance Monoid CubeSet where
  mempty = CubeSet 0 0 0

cubesAvailable :: CubeSet
cubesAvailable = CubeSet 12 13 14

maxCubeSets :: CubeSet -> CubeSet -> CubeSet
maxCubeSets (CubeSet n i j) (CubeSet x y z) =
  CubeSet (max n x) (max i y) (max j z)

main = do
  ls <- lines <$> readFile "input"
  let games = map parse ls
  putStr "Part 1: "
  print $ solve1 games

solve1 :: [[CubeSet]] -> Int
solve1 cubes = sum . map fst . filter (\(i, s) -> isWithinLimit s) $ zip [1..] sets
  where
    sets = map (foldl maxCubeSets mempty) cubes

isWithinLimit :: CubeSet -> Bool
isWithinLimit (CubeSet a b c) =
  a <= x && b <= y && c <= z
  where
    CubeSet x y z = cubesAvailable

parse :: String -> [CubeSet]
parse line = parseDraw . T.strip <$> T.splitOn (T.pack ";") usefulContent
  where
    usefulContent = T.strip $ T.pack $ tail $ dropWhile (/= ':') line

parseDraw :: T.Text -> CubeSet
parseDraw draw = cubeLists <&> parseCubeList <&> parseCube & mconcat
  where
    cubeLists = draw
          & T.splitOn (T.pack ",")
          <&> T.strip
          <&> T.unpack
    parseCubeList cube = case words cube of
      [num, color] -> (read num :: Int, color)
      _            -> error "Invalid format"

parseCube :: (Int, String) -> CubeSet
parseCube (n, "red")   = CubeSet n 0 0
parseCube (n, "green") = CubeSet 0 n 0
parseCube (n, "blue")  = CubeSet 0 0 n
parseCube _            = error "Invalid cube color"

File: day02/rs/src/main.rs
use std::fs;

#[derive(Debug)]
struct CubeSet {
    red: u32,
    green: u32,
    blue: u32,
}

fn main() {
    let cont = fs::read_to_string("input").expect("File not found");
    let lines: Vec<_> = cont.lines().collect();
    let games = parse_games(lines);
    println!("{:?}", games);
}

fn parse_games<'a>(lines: Vec<&'a str>) -> Vec<Vec<CubeSet>> {
    lines
        .iter()
        .map(|ln| {
            let useful_cont = ln.split(":").last().expect("Invalid format");
            let useful_cont = useful_cont.trim();
            let sets = useful_cont.split(";").map(|x| x.trim());

            sets.map(|set| {
                let mut cube_set = CubeSet {
                    red: 0,
                    green: 0,
                    blue: 0,
                };

                let string_sets = set.split(",").map(|x| x.trim());
                for set_str in string_sets {
                    parse_cube_set(&mut cube_set, set_str);
                }

                cube_set
            })
            .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>()
}

fn parse_cube_set<'a>(cube_set: &'a mut CubeSet, cube: &str) -> &'a CubeSet {
    match cube.split(" ").collect::<Vec<_>>().as_slice() {
        [n, color] => {
            let n = n.parse::<u32>().expect("Invalid amount");
            match *color {
                "red" => cube_set.red = n,
                "green" => cube_set.green = n,
                "blue" => cube_set.blue = n,
                _ => panic!("Invalid color"),
            }
        }
        _ => panic!("Invalid format"),
    }
    cube_set
}
